### ISSUE TITLE
feature request: Verify Messages Published to SNS Platform Endpoint

### ISSUE THREAD

**Issue by ShedPlant** [localstack/localstack#6360]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

I have an application which is responsible for push notifications. It publishes to SNS Platform Endpoints. I am trying to write acceptance tests for it, with localstack for mocked AWS services.

The application seems to be publishing successfully. However, is there any way for my test to subscribe, or get, the published messages out of localstack again?

My investigation thus far has found:
- https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#platformendpoint boto3 `PlatformEndpoint` resource does not have a `subscribe()` method, unlike `Topic`.
- https://github.com/localstack/localstack/issues/874
- https://github.com/localstack/localstack/pull/2818/files
- https://github.com/localstack/localstack/blob/9947a36f01b2eb60aea03cad1705348c3d17fe71/tests/integration/test_sns.py#L418 localstack has a `test_subscribe_platform_endpoint` which publishes to a SNS Platform Endpoint, and checks localstack's internal SNS Backend for the messages. But, I doubt I can get at this while running `localstack` as a docker image.
- https://github.com/localstack/localstack/pull/6256 added support for localstack to connect to real GCM, but I don't want to send to a real device, I just want to have automated tests that treat the application as a black box, as much as possible.

**Comment by salemcouja** [comment]
I'm also facing the same problem

**Comment by NickGrantX** [comment]
Can't wait to see @bentsku 's PR merged and released... I really need this!

**Comment by bentsku** [comment]
Hi @NickGrantX !
I'm going to give it a little push, I was currently not satisfied with the format of the response, I'll change that and hopefully merge it soon.
In the meantime, I'd love to hear about your use case:
With this PR, the messages will be available "as is", and will contain the same data that what you sent. Would this satisfy your use case?
I was thinking that we could maybe implement platform specifics in the future, depending on the targeted platform. This is not on the roadmap for now, and more of a future feature that could be part of a bigger refactoring of SNS. But I'm not sure there is an actual use case for this. That's why I'd be glad to hear more about yours! 😄

**Comment by NickGrantX** [comment]
>I'd love to hear about your use case

Sorry for the delay in replying @bentsku. We have a service which sends push notifications to target ARNs (the APNS/GCM targets) using the `Publish` API endpoint of the SNS service. For our integration testing we want to check that the AWS SNS API got called as expected.

Most useful for us, it to have a simple historical record of the API calls, in the sequence that they were received. We can figure everything out from that. In a perfect world, there would also be a "reset" endpoint too, which we could use to reset the history in-between tests.

FYI and for anyone else interested in this issue, here's a Python stub I wrote, which is our stop-gap hacky solution temporarily while we're waiting for a Localstack official solution:

```python
import json
import time
from http.server import BaseHTTPRequestHandler, HTTPServer
from threading import Thread

import requests


class SnsStub(BaseHTTPRequestHandler):
    """
    Stub AWS SNS API HTTP server, which keeps request history for testing purposes.

    This is our in-house developed AWS SNS stub/mock, because Localstack doesn't expose a good API to allow us to access
    the call history, and the other way we were doing it was quite unreliable/flakey.

    This stub can be deleted once this PR is merged and released: https://github.com/localstack/localstack/pull/6661
    """

    PORT = 7777
    HISTORY_PATH = '/history'
    request_history = []

    def do_POST(self):
        content_length = int(self.headers.get('Content-Length'))
        post_body = self.rfile.read(content_length)
        SnsStub.request_history.append({
            "path": self.path,
            "post_body": post_body.decode('utf8'),
        })

        response_content = """
                            <PublishResponse xmlns="https://sns.amazonaws.com/doc/2010-03-31/">
                                <PublishResult>
                                    <MessageId>567910cd-659e-55d4-8ccb-5aaf14679dc0</MessageId>
                                </PublishResult>
                                <ResponseMetadata>
                                    <RequestId>d74b8436-ae13-5ab4-a9ff-ce54dfea72a0</RequestId>
                                </ResponseMetadata>
                            </PublishResponse>
                            """
        self.send_response(requests.codes.accepted)
        self.send_header('Content-Type', 'text/xml;charset=UTF-8')
        self.end_headers()
        self.wfile.write(response_content.encode('utf-8'))

    def do_GET(self):
        if self.path != SnsStub.HISTORY_PATH:
            self.send_response_only(requests.codes.bad_request)
            self.end_headers()
            return

        self.send_response(requests.codes.ok)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.end_headers()
        response_content = json.dumps(SnsStub.request_history)
        self.wfile.write(response_content.encode('utf-8'))

    def do_DELETE(self):
        if self.path != SnsStub.HISTORY_PATH:
            self.send_response_only(requests.codes.bad_request)
            self.end_headers()
            return

        SnsStub.request_history = []
        self.send_response_only(requests.codes.ok)
        self.end_headers()


if __name__ == '__main__':
    mock_server = HTTPServer(('0.0.0.0', SnsStub.PORT), SnsStub)
    mock_server_thread = Thread(target=mock_server.serve_forever, daemon=True)
    mock_server_thread.start()

    print(f"Stub AWS SNS server is waiting to receive requests on port {SnsStub.PORT}...")
    while True:
        time.sleep(1)
```

**Comment by bentsku** [comment]
Hello @NickGrantX and thanks a lot for your detailed answer!
It seems the way I've done it for now should cover your use-case, perfect!
Sorry I have been busy with another project, I was also waiting on another PR to be merged to avoid conflict.
I will merge this next week, sorry for the delay as well about this! I think I will have time to add the `reset` endpoint then, that's a great idea, thanks!

**Comment by ShedPlant** [comment]
We are making use of this feature now, thanks @thrau , @bentsku and @DAtek 🍻 !

https://docs.localstack.cloud/user-guide/aws/sns/#accessing-published-platform-messages
